Ban "write" access for fields on continuous function spaces (for #138)

TeranIvy opened the pull request:
This PR removes `GH_WRITE` from the list of allowed accesses for fields on continuous function spaces. `ANY_SPACE` access for different looping types is dealt with in #471.

There are many test changes due to several test kernels using this access for continuous spaces. Some tranformation tests for parallel OMP DO loops were modified to use test examples on discontinuous spaces as it is not possible to do OMP on continuous spaces without colouring. Some of the modified test kernels had Fortran that could not be compiled so the code was corrected with generating subroutine body and enabling compilation in relevant tests.

TeranIvy commented:
Ready for review from @rupertford.

rupertford commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -5739,50 +5750,49 @@ def test_unexpected_type_error(dist_mem):
-                       create_arg_list.cma_operator]:
-            with pytest.raises(NotImplementedError):
-                method(None)
+def test_argordering_exceptions(dist_mem):
```
Great again.

rupertford commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -5990,25 +6000,27 @@ def test_anyw2_operators(dist_mem, tmpdir):
+                     distributed_memory=dist_mem).create(invoke_info)
+    generated_code = str(psy.gen)
+
+    assert LFRicBuild(tmpdir).code_compiles(psy)
```
Brill.

rupertford commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -6635,12 +6653,14 @@ def test_no_halo_exchange_annex_dofs(tmpdir, monkeypatch,
     else:
         assert "CALL f1_proxy%halo_exchange" in result
-    assert "CALL f2_proxy%halo_exchange" not in result
+        assert "CALL f2_proxy%halo_exchange" in result
```
Please add a check in `if annexed` that the call does not exist in the output.

TeranIvy replied:
Logical thing to do that I forgot :smiley: Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/loop_fuse_trans.py:
```diff
@@ -1,19 +1,56 @@
+# -----------------------------------------------------------------------------
```
Thanks for adding the missing copyright notice.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_anyw2_stencil_mod.f90:
```diff
@@ -24,21 +28,49 @@
 contains
-  subroutine testkern_anyw2_stencil_code()
+
+  subroutine testkern_anyw2_stencil_code(nlayers, fld1,                 &
```
Thanks for adding the arguments.

TeranIvy commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -344,15 +344,13 @@ def _validate_field(self, arg_type):
+        # TODO in issue #471: The kernels that loop over DoFs (built-ins)
+        # and are specified on 'ANY_SPACE_*' will have the same allowed
+        # accesses as discontinuous fields as they do not need colouring
+        # when updating quantities in parallel. This temporary accesses
```
Thanks for the suggestion, done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -373,6 +371,10 @@ def _validate_field(self, arg_type):
         # Check fields on continuous function spaces
         fld_cont_acc_msg = [rev_access_mapping[acc] for acc in
                             field_cont_accesses]
+        # As said above, this temporary access for messages will be
```
Done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -384,14 +386,14 @@ def _validate_field(self, arg_type):
         # As said above, allowed accesses for fields on ANY_SPACE depend on
         # type of looping. This will be refined in #471, however for now we
-        # use the accesses for continuous spaces (worst-case scenarios).
+        # use the temporary accesses for continuous spaces.
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -613,12 +615,20 @@ def test_check_seq_colours_omp_do(tmpdir, monkeypatch, annexed, dist_mem):
-    _, invoke = get_invoke("1_single_invoke.f90", TEST_API,
-                           name="invoke_0_testkern_type", dist_mem=dist_mem)
+    distributed memory is on or off.
+    Note: OpenMP parallel transformation can be applied on its own only on
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -613,12 +615,20 @@ def test_check_seq_colours_omp_do(tmpdir, monkeypatch, annexed, dist_mem):
+    distributed memory is on or off.
+    Note: OpenMP parallel transformation can be applied on its own only on
+    a discontinuous function space. However, colouring can be applied when
+    an argument has "INC" access and in LFRic API this is only possible
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -945,13 +971,12 @@ def test_loop_fuse_different_spaces(monkeypatch, dist_mem):
-            index = 8
+            index = 9
+            # f, c and g halo exchanges between loops can be moved
+            # before 1st loop as they are not accessed in first loop
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -4734,41 +4778,65 @@ def test_rc_continuous_halo_remove():
+    # of depth 1, as well as two halo exchanges of depth 1 in code
+    # (one before f3_inc_loop and one before the f3_read_loop).
+    assert result.count("CALL f3_proxy%halo_exchange(depth=1") == 2
+    assert "IF (f3_proxy%is_dirty(depth=1)) THEN" in result
```
Good idea, done!

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -4734,41 +4778,65 @@ def test_rc_continuous_halo_remove():
+    f3_read_hex = schedule.children[7]
+    f3_read_loop = schedule.children[9]
+    # f3 has "inc" access so there is a check for the halo exchange
+    # of depth 1, as well as two halo exchanges of depth 1 in code
```
I meant "two" for the f3 field. Modified the text as sugested and added clarification.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -4734,41 +4778,65 @@ def test_rc_continuous_halo_remove():
+    f3_read_loop = schedule.children[9]
+    # f3 has "inc" access so there is a check for the halo exchange
+    # of depth 1, as well as two halo exchanges of depth 1 in code
+    # (one before f3_inc_loop and one before the f3_read_loop).
```
I meant "one" for the f3 field. There are actually three halo exchanges before the f3_read_loop due to access changes in two kernels (see the comment above).

TeranIvy commented:
Addressed the review comments. Also wanted to update the branch with the master but @rupertford has already done that :smile: (thanks!) Tests with compilation pass. Ready for next review.
